### PR TITLE
fix(packagist): Replace V2 URL path instead of joining it

### LINKED ISSUE #20710: Gitlab private packagist repository has repeating URL
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Running via NPM (34.154.2)

### If you're self-hosting Renovate, select which platform you are using.

None

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I currently have self-hosted composer packages via Gitlab (which is the same instance as the repo). When running in debug mode, I can see it has part of the API url repeated.

Please note, I've replaced our Gitlab URL with `gitlab.selfhosted.com` to signify this is our self-hosted version

<details><summary><code>composer.json</code></summary>

We have the Gitlab composer json committed in the `composer.json` file - represented as `[COMPOSER_TOKEN]`. In the example below, `liquidlight/pavilion` is a private package

```json
{
    "name": "ll/basesite",
    "description": "Liquid Light Basesite",
    "type": "project",
    "require": {
        "liquidlight/pavilion": "^5.0"
    },
    "repositories": [
        {
            "type": "composer",
            "url": "https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/packages.json"
        }
    ],
    "config": {
        "allow-plugins": {
            "typo3/class-alias-loader": true,
            "typo3/cms-composer-installers": true,
            "helhum/dotenv-connector": true
        },
        "gitlab-domains": [
            "gitlab.selfhosted.com"
        ],
        "gitlab-token": {
            "gitlab.selfhosted.com": "[COMPOSER_TOKEN]"
        },
        "sort-packages": true
    },
    "extra": {
        "helhum/dotenv-connector": {
            "adapter": "Helhum\\DotEnvConnector\\Adapter\\SymfonyLoadEnv"
        },
        "typo3/cms": {
            "web-dir": "html"
        }
    },
}
```
</details>

<details><summary><code>config.js</code></summary>

```js
module.exports = {
	platform: 'gitlab',
	endpoint: 'https://gitlab.selfhosted.com/api/v4/',
	token: '[API_TOKEN]',

	gitAuthor: "RenovateBot <developers@liquidlight.co.uk>",

	repositories: [
		'll/ll/basesite',
	],

	requireConfig: 'required',
	onboarding: true,
	onboardingConfig: {
		extends: [
			"config:base",
			":preserveSemverRanges",
			":rebaseStalePrs",
			":enableVulnerabilityAlertsWithLabel('security')",
			"group:recommended"
		],
		commitMessageAction: 'Dep',
		prConcurrentLimit: 3,
		separateMinorPatch: false,
		major: {
			enabled: false
		},
		packageRules: [
			{
				matchManagers: ["composer"],
				rangeStrategy: "update-lockfile",
			},
		],
	},

	enabledManagers: [
		'composer',
		'npm',
	],
};

```
</details>

### Relevant debug logs

Logs are below, the issue is the repeating `api` section of the package URL

```
https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/api/v4/group/63/-/packages/composer/p2/liquidlight/pavilion.json
```

<details><summary>Logs</summary>

**Please ignore the npm errors** - They work with me adding a `npmrc` option to the `config.js` which I have excluded to ensure it wasn't interfering

```
> LOG_LEVEL=debug renovate --dry-run

DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in /Users/mike/Sites/rennovate-bot/config.js
 WARN: cli config dryRun property has been changed to full
DEBUG: File config
       "config": {
         "platform": "gitlab",
         "endpoint": "https://gitlab.selfhosted.com/api/v4/",
         "token": "***********",
         "gitAuthor": "RenovateBot <developers@liquidlight.co.uk>",
         "repositories": ["ll/ll/basesite"],
         "logLevel": "DEBUG",
         "requireConfig": "required",
         "onboarding": true,
         "onboardingConfig": {
           "extends": [
             "config:base",
             ":preserveSemverRanges",
             ":rebaseStalePrs",
             ":enableVulnerabilityAlertsWithLabel('security')",
             "group:recommended"
           ],
           "commitMessageAction": "Dep",
           "prConcurrentLimit": 3,
           "separateMinorPatch": false,
           "major": {"enabled": false},
           "packageRules": [
             {"matchManagers": ["composer"], "rangeStrategy": "update-lockfile"}
           ]
         },
         "enabledManagers": ["composer", "npm"]
       }
DEBUG: CLI config
       "config": {"dryRun": "full"}
DEBUG: Env config
       "config": {"hostRules": []}
DEBUG: Combined config
       "config": {
         "platform": "gitlab",
         "endpoint": "https://gitlab.selfhosted.com/api/v4/",
         "token": "***********",
         "gitAuthor": "RenovateBot <developers@liquidlight.co.uk>",
         "repositories": ["ll/ll/basesite"],
         "logLevel": "DEBUG",
         "requireConfig": "required",
         "onboarding": true,
         "onboardingConfig": {
           "extends": [
             "config:base",
             ":preserveSemverRanges",
             ":rebaseStalePrs",
             ":enableVulnerabilityAlertsWithLabel('security')",
             "group:recommended"
           ],
           "commitMessageAction": "Dep",
           "prConcurrentLimit": 3,
           "separateMinorPatch": false,
           "major": {"enabled": false},
           "packageRules": [
             {"matchManagers": ["composer"], "rangeStrategy": "update-lockfile"}
           ]
         },
         "enabledManagers": ["composer", "npm"],
         "hostRules": [],
         "dryRun": "full"
       }
DEBUG: Adding trailing slash to endpoint
DEBUG: Found valid git version: 2.37.1
DEBUG: GitLab version is: 15.9.1
DEBUG: Using configured gitAuthor (RenovateBot <developers@liquidlight.co.uk>)
DEBUG: Adding token authentication for gitlab.selfhosted.com to hostRules
DEBUG: Using baseDir: /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate
DEBUG: Using cacheDir: /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate/cache
DEBUG: Using containerbaseDir: /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate/cache/containerbase
DEBUG: Initializing Renovate internal cache into /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for gitlab.selfhosted.com to hostRules
DEBUG: validatePresets()
DEBUG: Reinitializing hostRules for repo
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for gitlab.selfhosted.com to hostRules
 INFO: Repository started (repository=ll/ll/basesite)
       "renovateVersion": "34.154.2"
DEBUG: Using localDir: /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate/repos/gitlab/ll/ll/basesite (repository=ll/ll/basesite)
DEBUG: PackageFiles.clear() - Package files deleted (repository=ll/ll/basesite)
DEBUG: ll/ll/basesite default branch = main (repository=ll/ll/basesite)
DEBUG: Enabling Git FS (repository=ll/ll/basesite)
DEBUG: Using http URL: https://gitlab.selfhosted.com/ll/ll/basesite.git (repository=ll/ll/basesite)
DEBUG: Resetting npmrc (repository=ll/ll/basesite)
DEBUG: checkOnboarding() (repository=ll/ll/basesite)
DEBUG: isOnboarded() (repository=ll/ll/basesite)
DEBUG: findFile(renovate.json) (repository=ll/ll/basesite)
DEBUG: Initializing git repository into /var/folders/1k/84znrvz53j1g1h68ctbbkvpr0000gn/T/renovate/repos/gitlab/ll/ll/basesite (repository=ll/ll/basesite)
DEBUG: Performing blobless clone (repository=ll/ll/basesite)
DEBUG: git clone completed (repository=ll/ll/basesite)
       "durationMs": 1557
DEBUG: latest repository commit (repository=ll/ll/basesite)
       "latestCommit": {
         "hash": "5e24391b0520e5d23e1ad7cc308671040fc7c5e1",
         "date": "2023-03-02T10:42:15+00:00",
         "message": "Chore: Remove PHP dependency",
         "refs": "HEAD -> main, origin/main, origin/HEAD",
         "body": "",
         "author_name": "Mike Street",
         "author_email": "mike@liquidlight.co.uk"
       }
DEBUG: findFile(renovate.json5) (repository=ll/ll/basesite)
DEBUG: findFile(.github/renovate.json) (repository=ll/ll/basesite)
DEBUG: findFile(.github/renovate.json5) (repository=ll/ll/basesite)
DEBUG: findFile(.gitlab/renovate.json) (repository=ll/ll/basesite)
DEBUG: findFile(.gitlab/renovate.json5) (repository=ll/ll/basesite)
DEBUG: findFile(.renovaterc) (repository=ll/ll/basesite)
DEBUG: findFile(.renovaterc.json) (repository=ll/ll/basesite)
DEBUG: findFile(.renovaterc.json5) (repository=ll/ll/basesite)
DEBUG: config file not found (repository=ll/ll/basesite)
DEBUG: findPr(renovate/configure, Configure Renovate, !open) (repository=ll/ll/basesite)
DEBUG: Found no closed onboarding PR (repository=ll/ll/basesite)
DEBUG: Repo is not onboarded (repository=ll/ll/basesite)
DEBUG: getBranchPr(renovate/configure) (repository=ll/ll/basesite)
DEBUG: findPr(renovate/configure, undefined, open) (repository=ll/ll/basesite)
DEBUG: getPr(3) (repository=ll/ll/basesite)
DEBUG: getMR(3) (repository=ll/ll/basesite)
DEBUG: Onboarding PR already exists (repository=ll/ll/basesite)
DEBUG: Checking if onboarding branch needs rebasing (repository=ll/ll/basesite)
DEBUG: Checking if this org/owner has a default Renovate preset which can be used. (repository=ll/ll/basesite)
DEBUG: GET https://gitlab.selfhosted.com/api/v4/projects/ll%2Frenovate-config = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=77) (repository=ll/ll/basesite)
DEBUG: GitLab API 404 (repository=ll/ll/basesite)
       "url": "https://gitlab.selfhosted.com/api/v4/projects/ll%2Frenovate-config"
DEBUG: Preset file default.json not found in ll/renovate-config (repository=ll/ll/basesite)
DEBUG: GitLab API 404 (repository=ll/ll/basesite)
       "url": "https://gitlab.selfhosted.com/api/v4/projects/ll%2Frenovate-config"
DEBUG: Preset file renovate.json not found in ll/renovate-config (repository=ll/ll/basesite)
DEBUG: GET https://gitlab.selfhosted.com/api/v4/projects/ll%2F.gitlab = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=53) (repository=ll/ll/basesite)
DEBUG: GitLab API 404 (repository=ll/ll/basesite)
       "url": "https://gitlab.selfhosted.com/api/v4/projects/ll%2F.gitlab"
DEBUG: Preset file renovate-config.json not found in ll/.gitlab (repository=ll/ll/basesite)
DEBUG: No default org/owner preset found, so the default onboarding config will be used instead. Note: do not be concerned with any 404 messages that preceded this. (repository=ll/ll/basesite)
DEBUG: onboarding config (repository=ll/ll/basesite)
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": [
           "config:base",
           ":preserveSemverRanges",
           ":rebaseStalePrs",
           ":enableVulnerabilityAlertsWithLabel('security')",
           "group:recommended"
         ],
         "commitMessageAction": "Dep",
         "prConcurrentLimit": 3,
         "separateMinorPatch": false,
         "major": {"enabled": false},
         "packageRules": [
           {"matchManagers": ["composer"], "rangeStrategy": "update-lockfile"}
         ]
       }
DEBUG: branch.isModified(): using git to calculate (repository=ll/ll/basesite)
DEBUG: branch.isModified() = false (repository=ll/ll/basesite)
DEBUG: setCachedModifiedResult(): Branch cache not present (repository=ll/ll/basesite)
DEBUG: Rebasing onboarding branch (repository=ll/ll/basesite)
 INFO: DRY-RUN: Would rebase files in onboarding branch (repository=ll/ll/basesite)
DEBUG: No renovate config file found (repository=ll/ll/basesite)
DEBUG: migrateAndValidate() (repository=ll/ll/basesite)
DEBUG: No config migration necessary (repository=ll/ll/basesite)
DEBUG: massaged config (repository=ll/ll/basesite)
       "config": {}
DEBUG: migrated config (repository=ll/ll/basesite)
       "config": {}
DEBUG: Found repo ignorePaths (repository=ll/ll/basesite)
       "ignorePaths": ["**/node_modules/**", "**/bower_components/**"]
DEBUG: No vulnerability alerts found (repository=ll/ll/basesite)
DEBUG: No baseBranches (repository=ll/ll/basesite)
DEBUG: extract() (repository=ll/ll/basesite)
DEBUG: Setting current branch to main (repository=ll/ll/basesite)
DEBUG: latest commit (repository=ll/ll/basesite)
       "branchName": "main",
       "latestCommitDate": "2023-03-02T10:42:15+00:00"
DEBUG: Applying enabledManagers filtering (repository=ll/ll/basesite)
DEBUG: Using file match: (^|/)([\w-]*)composer\.json$ for manager composer (repository=ll/ll/basesite)
DEBUG: Using file match: (^|/)package\.json$ for manager npm (repository=ll/ll/basesite)
DEBUG: Matched 2 file(s) for manager composer: app/basesite/composer.json, composer.json (repository=ll/ll/basesite)
DEBUG: Matched 1 file(s) for manager npm: package.json (repository=ll/ll/basesite)
DEBUG: npm file package.json has name "@ll/basesite" (repository=ll/ll/basesite)
DEBUG: Found composer lock file composer.json (repository=ll/ll/basesite)
DEBUG: Detecting pnpm Workspaces (repository=ll/ll/basesite)
DEBUG: Detecting Lerna and Yarn Workspaces (repository=ll/ll/basesite)
DEBUG: Finding locked versions (repository=ll/ll/basesite)
DEBUG: Found package-lock.json for package.json (repository=ll/ll/basesite)
DEBUG: Found composer package files (repository=ll/ll/basesite)
DEBUG: Found npm package files (repository=ll/ll/basesite)
DEBUG: Found 3 package file(s) (repository=ll/ll/basesite)
 INFO: Dependency extraction complete (repository=ll/ll/basesite, baseBranch=main)
       "stats": {
         "managers": {
           "composer": {"fileCount": 2, "depCount": 3},
           "npm": {"fileCount": 1, "depCount": 6}
         },
         "total": {"fileCount": 3, "depCount": 9}
       }
DEBUG: Dependency app/basesite has unsupported value @dev (repository=ll/ll/basesite)
DEBUG: GET https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/api/v4/group/63/-/packages/composer/p2/liquidlight/pavilion.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=66) (repository=ll/ll/basesite)
DEBUG: GET https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/api/v4/group/63/-/packages/composer/p2/liquidlight/pavilion~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=134) (repository=ll/ll/basesite)
DEBUG: GET https://packagist.org/p2/liquidlight/pavilion.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=120) (repository=ll/ll/basesite)
DEBUG: Datasource 404 (repository=ll/ll/basesite)
       "datasource": "packagist",
       "packageName": "liquidlight/pavilion",
       "url": "https://packagist.org/p2/liquidlight/pavilion.json"
DEBUG: Failed to look up packagist dependency liquidlight/pavilion (repository=ll/ll/basesite, packageFile=composer.json, dependency=liquidlight/pavilion)
DEBUG: GET https://packagist.org/p2/liquidlight/pavilion~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=122) (repository=ll/ll/basesite)
DEBUG: GET https://registry.npmjs.org/@packages%2Fcustom-fonts = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=256) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/custom-fonts (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/custom-fonts)
DEBUG: GET https://registry.npmjs.org/@packages%2Ftabs = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=254) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/tabs (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/tabs)
DEBUG: GET https://registry.npmjs.org/@packages%2Fgulp = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=257) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/gulp (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/gulp)
DEBUG: GET https://registry.npmjs.org/@packages%2Ftoggle = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=205) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/toggle (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/toggle)
DEBUG: GET https://registry.npmjs.org/@packages%2Fservice-worker = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=683) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/service-worker (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/service-worker)
DEBUG: GET https://registry.npmjs.org/@packages%2Fnavigation = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=701) (repository=ll/ll/basesite)
DEBUG: Failed to look up npm dependency @packages/navigation (repository=ll/ll/basesite, packageFile=package.json, dependency=@packages/navigation)
DEBUG: PackageFiles.add() - Package file saved for base branch (repository=ll/ll/basesite, baseBranch=main)
DEBUG: Package releases lookups complete (repository=ll/ll/basesite, baseBranch=main)
DEBUG: branchifyUpgrades (repository=ll/ll/basesite)
DEBUG: detectSemanticCommits() (repository=ll/ll/basesite)
DEBUG: getCommitMessages (repository=ll/ll/basesite)
DEBUG: semanticCommits: detected "jquery" (repository=ll/ll/basesite)
DEBUG: semanticCommits: disabled (repository=ll/ll/basesite)
DEBUG: 1 flattened updates found: typo3/cms-core (repository=ll/ll/basesite)
DEBUG: Returning 1 branch(es) (repository=ll/ll/basesite)
DEBUG: config.repoIsOnboarded=false (repository=ll/ll/basesite)
DEBUG: packageFiles with updates (repository=ll/ll/basesite, baseBranch=main)
       "config": {
         "composer": [
           {
             "deps": [
               {
                 "depType": "require",
                 "depName": "typo3/cms-core",
                 "currentValue": "^10.4",
                 "datasource": "packagist",
                 "updates": [
                   {
                     "bucket": "major",
                     "newVersion": "12.2.0",
                     "newValue": "^12.0",
                     "releaseTimestamp": "2023-02-07T10:24:25.000Z",
                     "newMajor": 12,
                     "newMinor": 2,
                     "updateType": "major",
                     "isRange": true,
                     "branchName": "renovate/typo3-cms-core-12.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "composer",
                 "sourceUrl": "https://github.com/TYPO3-CMS/core",
                 "registryUrl": "https://packagist.org",
                 "homepage": "https://typo3.org",
                 "currentVersion": "10.4.36",
                 "isSingleVersion": false
               }
             ],
             "managerData": {"composerJsonType": "typo3-cms-extension"},
             "packageFile": "app/basesite/composer.json"
           },
           {
             "deps": [
               {
                 "depType": "require",
                 "depName": "app/basesite",
                 "currentValue": "@dev",
                 "datasource": "packagist",
                 "registryUrls": [
                   "https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer",
                   "https://packagist.org"
                 ],
                 "updates": [],
                 "warnings": [],
                 "versioning": "composer",
                 "skipReason": "invalid-value"
               },
               {
                 "depType": "require",
                 "depName": "liquidlight/pavilion",
                 "currentValue": "^5.0",
                 "datasource": "packagist",
                 "lockedVersion": "5.7.0",
                 "registryUrls": [
                   "https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer",
                   "https://packagist.org"
                 ],
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "liquidlight/pavilion",
                     "message": "Failed to look up packagist dependency liquidlight/pavilion"
                   }
                 ],
                 "versioning": "composer"
               }
             ],
             "lockFiles": ["composer.lock"],
             "managerData": {"composerJsonType": "project"},
             "packageFile": "composer.json"
           }
         ],
         "npm": [
           {
             "deps": [
               {
                 "depType": "dependencies",
                 "depName": "@packages/custom-fonts",
                 "currentValue": "^1.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/custom-fonts",
                     "message": "Failed to look up npm dependency @packages/custom-fonts"
                   }
                 ],
                 "versioning": "npm"
               },
               {
                 "depType": "dependencies",
                 "depName": "@packages/gulp",
                 "currentValue": "^3.0.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/gulp",
                     "message": "Failed to look up npm dependency @packages/gulp"
                   }
                 ],
                 "versioning": "npm"
               },
               {
                 "depType": "dependencies",
                 "depName": "@packages/navigation",
                 "currentValue": "^1.0.1",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/navigation",
                     "message": "Failed to look up npm dependency @packages/navigation"
                   }
                 ],
                 "versioning": "npm"
               },
               {
                 "depType": "dependencies",
                 "depName": "@packages/service-worker",
                 "currentValue": "^1.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/service-worker",
                     "message": "Failed to look up npm dependency @packages/service-worker"
                   }
                 ],
                 "versioning": "npm"
               },
               {
                 "depType": "dependencies",
                 "depName": "@packages/tabs",
                 "currentValue": "^2.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/tabs",
                     "message": "Failed to look up npm dependency @packages/tabs"
                   }
                 ],
                 "versioning": "npm"
               },
               {
                 "depType": "dependencies",
                 "depName": "@packages/toggle",
                 "currentValue": "^1.0",
                 "datasource": "npm",
                 "prettyDepType": "dependency",
                 "lockedVersion": null,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "@packages/toggle",
                     "message": "Failed to look up npm dependency @packages/toggle"
                   }
                 ],
                 "versioning": "npm"
               }
             ],
             "packageFileVersion": "1.0.0",
             "managerData": {
               "npmLock": "package-lock.json",
               "packageJsonName": "@ll/basesite",
               "yarnZeroInstall": false,
               "hasPackageManager": false
             },
             "skipInstalls": true,
             "packageFile": "package.json",
             "lockFiles": ["package-lock.json"],
             "constraints": {}
           }
         ]
       }
DEBUG: detectSemanticCommits() (repository=ll/ll/basesite)
DEBUG: semanticCommits: returning "disabled" from cache (repository=ll/ll/basesite)
DEBUG: ensureOnboardingPr() (repository=ll/ll/basesite)
DEBUG: getBranchPr(renovate/configure) (repository=ll/ll/basesite)
DEBUG: findPr(renovate/configure, undefined, open) (repository=ll/ll/basesite)
DEBUG: getPr(3) (repository=ll/ll/basesite)
DEBUG: getMR(3) (repository=ll/ll/basesite)
DEBUG: Filling in onboarding PR template (repository=ll/ll/basesite)
 INFO: DRY-RUN: Would check branch renovate/configure (repository=ll/ll/basesite)
DEBUG: Found package lookup warnings (repository=ll/ll/basesite)
       "warnings": [
         "Failed to look up packagist dependency liquidlight/pavilion",
         "Failed to look up npm dependency @packages/custom-fonts",
         "Failed to look up npm dependency @packages/gulp",
         "Failed to look up npm dependency @packages/navigation",
         "Failed to look up npm dependency @packages/service-worker",
         "Failed to look up npm dependency @packages/tabs",
         "Failed to look up npm dependency @packages/toggle"
       ],
       "warningFiles": ["composer.json", "package.json"]
DEBUG: getPrList() (repository=ll/ll/basesite)
DEBUG: Found open onboarding PR (repository=ll/ll/basesite)
 INFO: DRY-RUN: Would update onboarding PR (repository=ll/ll/basesite)
DEBUG: processRepo() (repository=ll/ll/basesite)
DEBUG: getPrCache() (repository=ll/ll/basesite)
 INFO: DRY-RUN: Would close Dependency Dashboard (repository=ll/ll/basesite)
       "title": "Dependency Dashboard"
 INFO: DRY-RUN: Would save repository cache. (repository=ll/ll/basesite)
DEBUG: Removing any stale branches (repository=ll/ll/basesite)
DEBUG: config.repoIsOnboarded=false (repository=ll/ll/basesite)
DEBUG: Branch lists (repository=ll/ll/basesite)
       "branchList": ["renovate/configure"],
       "renovateBranches": ["renovate/configure"]
DEBUG: remainingBranches= (repository=ll/ll/basesite)
DEBUG: No branches to clean up (repository=ll/ll/basesite)
DEBUG: ensureIssueClosing() (repository=ll/ll/basesite)
DEBUG: PackageFiles.clear() - Package files deleted (repository=ll/ll/basesite)
DEBUG: Repo is activated (repository=ll/ll/basesite)
DEBUG: Branch summary (repository=ll/ll/basesite)
       "cacheModified": undefined,
       "baseBranches": [{"branchName": "main", "sha": "5e24391b0520e5d23e1ad7cc308671040fc7c5e1"}],
       "branches": [],
       "inactiveBranches": ["renovate/typo3-cms-core-12.x"]
DEBUG: Renovate repository PR statistics (repository=ll/ll/basesite)
       "stats": {"total": 1, "open": 0, "closed": 0, "merged": 1}
DEBUG: Repository result: undefined, status: activated, enabled: true, onboarded: true (repository=ll/ll/basesite)
DEBUG: Repository timing splits (milliseconds) (repository=ll/ll/basesite)
       "splits": {"init": 3377, "extract": 92, "lookup": 789, "onboarding": 7, "update": 0},
       "total": 4394
DEBUG: Package cache statistics (repository=ll/ll/basesite)
       "get": {"count": 11, "avgMs": 4, "medianMs": 4, "maxMs": 12},
       "set": {"count": 1, "avgMs": 25}
DEBUG: http statistics (repository=ll/ll/basesite)
       "urls": {
         "https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/api/v4/group/63/-/packages/composer/p2/liquidlight/pavilion.json (GET,404)": 1,
         "https://gitlab.selfhosted.com/api/v4/group/63/-/packages/composer/api/v4/group/63/-/packages/composer/p2/liquidlight/pavilion~dev.json (GET,404)": 1,
         "https://gitlab.selfhosted.com/api/v4/projects/ll%2F.gitlab (GET,404)": 1,
         "https://gitlab.selfhosted.com/api/v4/projects/ll%2Fll%2Fbasesite/issues (GET,200)": 1,
         "https://gitlab.selfhosted.com/api/v4/projects/ll%2Fll%2Fbasesite/merge_requests (GET,200)": 1,
         "https://gitlab.selfhosted.com/api/v4/projects/ll%2Fll%2Fbasesite/merge_requests/3 (GET,200)": 1,
         "https://gitlab.selfhosted.com/api/v4/projects/ll%2Frenovate-config (GET,404)": 1,
         "https://packagist.org/p2/liquidlight/pavilion.json (GET,404)": 1,
         "https://packagist.org/p2/liquidlight/pavilion~dev.json (GET,404)": 1,
         "https://packagist.org/p2/typo3/cms-core.json (GET,200)": 1,
         "https://packagist.org/p2/typo3/cms-core~dev.json (GET,200)": 1,
         "https://registry.npmjs.org/@packages%2Fcustom-fonts (GET,404)": 1,
         "https://registry.npmjs.org/@packages%2Fgulp (GET,404)": 1,
         "https://registry.npmjs.org/@packages%2Fnavigation (GET,404)": 1,
         "https://registry.npmjs.org/@packages%2Fservice-worker (GET,404)": 1,
         "https://registry.npmjs.org/@packages%2Ftabs (GET,404)": 1,
         "https://registry.npmjs.org/@packages%2Ftoggle (GET,404)": 1
       },
       "hostStats": {
         "gitlab.selfhosted.com": {
           "requestCount": 7,
           "requestAvgMs": 113,
           "queueAvgMs": 0
         },
         "packagist.org": {"requestCount": 4, "requestAvgMs": 141, "queueAvgMs": 0},
         "registry.npmjs.org": {
           "requestCount": 6,
           "requestAvgMs": 393,
           "queueAvgMs": 0
         }
       },
       "totalRequests": 17
DEBUG: dns cache (repository=ll/ll/basesite)
       "hosts": []
 INFO: Repository finished (repository=ll/ll/basesite)
       "cloned": true,
       "durationMs": 4394
DEBUG: Renovate exiting
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Introduce `replaceUrlPath()` helper that is similar to `resolveBaseUrl`, but replaces the path of `baseUrl` instead of joining with it
- Make sure Packagist datasource uses it when required

## Context

- Closes: #20710

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Packagist test possible?

**Comment by zharinov** [comment]
Yeah, testing it...

**Comment by zharinov** [comment]
Test: https://github.com/renovate-testing/test-19495-composer-v2-4/pull/1

File: lib/util/url.spec.ts [hunk 1]
```diff
@@ -6,57 +6,91 @@ import {
...
+    ${'http://foo.io/'}     | ${'/aaa'}               | ${'http://foo.io/aaa'}
+    ${'http://foo.io/'}     | ${'aaa'}                | ${'http://foo.io/aaa'}
+    ${'http://foo.io'}      | ${'/aaa/'}              | ${'http://foo.io/aaa/'}
+    ${'http://foo.io'}      | ${'aaa/'}               | ${'http://foo.io/aaa/'}
+    ${'http://foo.io/'}     | ${'/aaa/'}              | ${'http://foo.io/aaa/'}
+    ${'http://foo.io/'}     | ${'aaa/'}               | ${'http://foo.io/aaa/'}
+    ${'http://foo.io/aaa'}  | ${'/bbb'}               | ${'http://foo.io/bbb'}
+    ${'http://foo.io/aaa'}  | ${'bbb'}                | ${'http://foo.io/bbb'}
+    ${'http://foo.io/aaa/'} | ${'/bbb'}               | ${'http://foo.io/bbb'}
+    ${'http://foo.io/aaa/'} | ${'bbb'}                | ${'http://foo.io/bbb'}
```
Comment (viceice) [inline lib/util/url.spec.ts:72]: 🤔 i would expect this to be `http://foo.io/aaa/bbb`

Comment (zharinov) [reply lib/util/url.spec.ts:72]: It's `replaceUrlPath`, not `resolveBaseUrl`

**Review by viceice — APPROVED** [review]
🤷‍♂️ both function looks broken to me.

we should create a proper url resolve function which fully works against official url Specs.
the we should validate registry url inside datasources to have proper format, eg no API path and trailing slash.

**Pull request merged**